### DISCUSSION TITLE
micropython feature suggestions

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#9865] [Ideas]
Is there a mechanism for suggesting changes to upython for those of us lacking the wherewithal to do pull requests? For example I really liked the nvs_set & nvs_get functionality of pycom python, &  that rtc memory preserved those values through a reset. Seemed to be a good alternative to rtc.memory() for multiple variables that need to be persevered through a deepsleep.

**Comment by jimmo** [comment]
This is the exact place to do that!

But in this case, NVS is actually supported. See https://docs.micropython.org/en/latest/library/esp32.html?highlight=nvs#esp32.NVS for docs and https://github.com/micropython/micropython/blob/master/tests/esp32/esp32_nvs.py for examples of how to use it.

**Reply by kjm1102** [reply]
I tried an 
```
import NVS
```
in a 1.17.0 release esp32 but got an import error, so I guess I need to download 1.19.0? Also do you know if the tweak to preserve rtc memory through a reset was added at the same time as the NVS?

**Reply by jimmo** [reply]
NVS was added way back in in v1.15. Can you show an example of getting ImportError?

**Reply by kjm1102** [reply]
>>> import os; os.uname() 
(sysname='esp32', nodename='esp32', release='1.17.0', version='v1.17 on 2021-09-02', machine='ESP32 module with ESP32')
>>> import NVS
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: no module named 'NVS'

**Reply by jimmo** [reply]
See the example I linked to:

```
from esp32 import NVS
```

`NVS` is the class, `esp32` is the module that you import it from.

**Reply by kjm1102** [reply]
tnx Jim so is 
```
nvs = NVS("mp-test")
```
the preferred way of initialising or should I substitute something else for "mp-test" (nvs=NVS() gives a TypeError wanting a positional argument)?

**Reply by jimmo** [reply]
It's the namespace name, you can choose whatever you want, just as long as you choose something. You can also have more than one namespace if you want to have different sets of key/values for different purposes. Think of it like a very simple one-layer directory structure:

https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/storage/nvs_flash.html

**Reply by kjm1102** [reply]
Understood. 

How do I recover a string stored as a blob? The example shows how to recover the strings length but not the string itself? My best guess fails.
```
from esp32 import NVS
nvs=NVS('locals')
blob1 = "testing a string as a blob"
nvs.set_blob("blob1", blob1)
print(nvs.get_blob("blob1"))
Traceback (most recent call last):
  File "<stdin>", line 8, in <module>
TypeError: function takes 3 positional arguments but 2 were given
```

**Reply by karfas** [reply]
Is something wrong with https://docs.micropython.org/en/v1.15/library/esp32.html ?
This reads
> `NVS.get_blob(key, buffer)`

There is no `buffer` argument in your get_blob() call.

**Comment by karfas** [comment]
@kjm1102: You are mixing things up.
The NVS class provides a key-value store in EEPROM in a separate partition. This is also used by some ESP-IDF functions.
RTC.memory() gives you access to a small portion of RTC memory (RAM).

Regarding RTC memory there are two of my PRs (and possibly others) open:
- (April '21) #7133 
- (Mai '22) #7298 

Additional, you could try https://github.com/karfas/upy-esp32-lib - there is some python code for managing RTC memory content.

**Reply by jimmo** [reply]
Yes that's a good point.

@kjm1102 RTC will not persist across a power cycle (unless you have an RTC backup battery) whereas NVS will, as it's written to flash (which I guess counts as EEPROM -- either way it's the same flash that also stores the firmware and the MicroPython filesystem). Either way I just assumed that NVS is what you want so that's what I described.

**Reply by kjm1102** [reply]
I thought NVS used rtc memory, my mistake. If it uses flash am I right in thinking it could cause wear problems long term the same as file write reportedly does?

**Reply by jimmo** [reply]
See the Espressif FAQ:
https://espressif-docs.readthedocs-hosted.com/projects/espressif-esp-faq/en/latest/software-framework/storage/nvs.html

**Reply by kjm1102** [reply]
Tnx.

Still can't figure out how to get a string stored as a blob, all the docs seem focussed on getting the length?

**Reply by jimmo** [reply]
It _returns_ the length, but stores the blob in the specified buffer. See the examples in the file I linked to https://github.com/micropython/micropython/blob/master/tests/esp32/esp32_nvs.py#L18

**Reply by kjm1102** [reply]
Nope don't get it. Nowhere in there can I see anything stored with nvs.set_blob recovered with nvs.get_blob. The _i32 example is fine integer stored, integer recovered, no problems.

**Reply by jimmo** [reply]
```py
buf1 = bytearray(len(blob1))
len1 = nvs.get_blob("blob1", buf1)
print(buf1)
```

this gets the blob, storing it in buf1 (which is then printed). very similar to readinto (used in socket/file/i2c).

**Reply by kjm1102** [reply]
Ah buf1 doing double duty, and buf1.decode() has the string, finally! The pycom version seemed a bit more intuitive but then what would I know.

What's with the nvs.commit() window dressing? Seems to work OK without it?

**Reply by karfas** [reply]
Can you please **read** the documentation ?
The documentation (see link above) clearly states
> Changes to NVS need to be committed to flash by calling the commit method. Failure to call commit results in changes being lost at the next reset.

**Reply by jimmo** [reply]
> The pycom version seemed a bit more intuitive but then what would I know.

It's done this way for efficiency, saves allocating a buffer unnecessarily for the result, instead you reuse a preallocated buffer. Same as readinto for socket/file/i2c.

**Comment by rkompass** [comment]
Referring to the title **_micropython feature suggestions_** (in plural) and encouraged by @jimmo s **_This is the exact place to do that!_** I would like to express some thoughts regarding the development process. (I consider myself also as lacking the wherewithal to do pull requests atm, so I rather think I'm not in a position to do that, but it circulates in my mind anyway.)

1. My impression is and you (@jimmo) stated that also explicitly, that you have a very very long todo list.
2. About every two weeks or so a new architecture comes into play. This of course is a result of micropythons huge success.
3. Also at least from the appearance of esp8266 onwards the development has to deal more and more with topics like networking, bluetooth, threading, webserver etc, which are important for the internet of things but may be considered a bit external from the pure python, native, viper, assembler, interrupt and common api for peripherals stuff. Many ideas concerning the latter are found in the early discussions but apparently there was no time to work on them.
4. The number of **core** developers at the same time seems to keep rather constant. I explain this to myself by the complexity of uPy s internals, which are not documented (you have to explore the sources to get a grip of it).

So I'm wondering whether it would be a good idea to collect and sort out the topics on the todo list from time to time and then start a kickstarter-like process to have an incentive (and focus?) for core development. I would like to support such a thing. But perhaps the approaches to different issues have to 'get ripe' before any real coding should start and it would be contraproductive to try to push it?
Probably you have discussed this from time to time. I would be glad to get an impression what the position of core developers is related to this.

**Reply by jimmo** [reply]
Thanks @rkompass -- you raise some excellent points. You're definitely right that a huge amount of our effort actually has very little to do with the core "Python" part.

As far as making the todo list a bit more open and clearer, that's definitely something I'm keen to do (if nothing else than to force us to focus too!). We talked a little bit about some of this stuff on the CircuitPython Show interview recently. And you're also right that it's very difficult to figure out how something should work until you actually sit down to code it -- this is also one of the things that makes reviewing PRs so much work.

I'm not sure Damien would ever want to do another Kickstarter, but GitHub sponsorship has definitely been a huge help in this area.
